File: src/Array/python/haoxiang/UniquePaths2.py


class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        # line = len(obstacleGrid)
        # row = len(obstacleGrid[0])
        #
        # res = [0]*(row+1)
        # for i in range(1,line):
        #     for j in range(1,row):
        #         res[j] = res[j]+res[j-1]  if obstacleGrid[i-1][j-1]==0 else 0
        # return res[line-1]
        # use DP
        sumNum = 0
        line = len(obstacleGrid)
        row = len(obstacleGrid[0])
        res= [[0 for x in range(row)] for x in range(line)]

        # if (m<2 or n<2):
        #     return 1
        # if (m==n==2):
        #     return 2
        for x in range(line):
            res[x][0] = 1
        for y in range(row):
            res[0][y] = 1
        for i in range(1,line):
            for j in range(1,row):
                res[i][j] = res[i-1][j]+res[i][j-1] if obstacleGrid[i-1][j-1]==0 else 0
        return res[line-1][row-1]

Test = Solution()

File: src/qiuyi/Array/UniquePaths-2.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        vector<vector<int>> cache(m, vector<int>(n, 0));

        for (int j = n-1; j >= 0; j--)
            if (obstacleGrid[m-1][j] != 1)
                cache[m-1][j] = 1;
            else
                break;
        for (int i = m-1; i >= 0; i--)
            if (obstacleGrid[i][n-1] != 1)
                cache[i][n-1] = 1;
            else
                break;
        for (int i = m-2; i >= 0; i--)
            for (int j = n-2; j >= 0; j--)
                if (obstacleGrid[i][j] != 1)
                    cache[i][j] = cache[i+1][j] + cache[i][j+1];
                else
                    continue;
        return cache[0][0];
    }
};